x/shield/keeper/purchase.go

```diff
@@ -77,54 +77,61 @@ func (k Keeper) DequeuePurchase(ctx sdk.Context, purchaseList types.PurchaseList
 }
 
 // PurchaseShield purchases shield of a pool.
-func (k Keeper) PurchaseShield(ctx sdk.Context, poolID uint64, shield sdk.Coins, description string, purchaser sdk.AccAddress) (types.Purchase, error) {
+func (k Keeper) PurchaseShield(ctx sdk.Context, poolID uint64, shield sdk.Coins, description string, purchaser sdk.AccAddress, serviceFees sdk.Coins) (types.Purchase, error) {
 	pool, found := k.GetPool(ctx, poolID)
 	if !found {
 		return types.Purchase{}, types.ErrNoPoolFound
 	}
-	poolParams := k.GetPoolParams(ctx)
-	claimParams := k.GetClaimProposalParams(ctx)
-
-	// check preconditions
-	if err := k.PoolShieldAvailable(ctx, pool); err != nil {
-		return types.Purchase{}, err
+	if !pool.Active {
+		return types.Purchase{}, types.ErrPoolInactive
 	}
 
+	// Check available collaterals.
 	shieldAmt := shield.AmountOf(k.sk.BondDenom(ctx))
-	if shieldAmt.GT(pool.Available) {
-		return types.Purchase{}, types.ErrNotEnoughShield
+	totalCollateral := k.GetTotalCollateral(ctx)
+	totalWithdrawing := k.GetTotalWithdrawing(ctx)
+	totalShield := k.GetTotalShield(ctx)
+	if totalShield.Add(shieldAmt).GT(totalCollateral.Sub(totalWithdrawing)) {
+		return types.Purchase{}, types.ErrNotEnoughCollateral
 	}
-	if pool.SponsorAddr.Equals(purchaser) {
-		return types.Purchase{}, types.ErrSponsorPurchase
+
+	// Check pool shield limit.
+	poolParams := k.GetPoolParams(ctx)
+	maxShield := totalCollateral.Sub(totalWithdrawing).ToDec().Mul(poolParams.PoolShieldLimit).TruncateInt()
+	if shieldAmt.GT(maxShield) {
```

Comment:
should `maxShield` be compared against `shieldAmt` + `pool.Shield`?

Reply:
Right it should be fixed.